# 파이썬입문 겨울방학 특강 (2)

`-` 오늘오전: 자료형, 오브젝트 .. 

`-` 오늘오후: 참조, 얕은복사, 깊은복사 

`-` 내일오전: 뷰와 카피 

`-` 내일오후: 개발환경 비교 

### 참조와 에일리어싱 

`-` 아래의 코드를 관찰하자. 

In [1]:
a=[1,2,3]
b=a
a=a+[4]

현재 a,b의 출력결과는? 

In [6]:
a,b

([1, 2, 3, 4], [1, 2, 3])

`-` 이제 다시 아래의 코드를 관찰하자. 

In [7]:
a=[1,2,3]
b=a
a.append(4) 

In [8]:
a,b

([1, 2, 3, 4], [1, 2, 3, 4])

`-` 아래의 코드를 다시 살펴보자. 

```python
a=[1,2,3]
b=a
a.append(4)
```

a,b라는 변수들은 메모리에 어떻게 저장이 되어있을까? 

상상력을 조금 발휘하면 아래와 같이 여길 수 있다. 

(1) 메모리는 변수를 담을 방이 여러개 있는 호텔이라고 생각하자. 

(2) 아래를 실행하였을 경우 
```python
a=[1,2,3]
```
- 메모리주소1에 존재하는 방을 a라고 하고, 그 방에 [1,2,3]을 넣는다. 

(3) 아래를 실행하였을 경우 
```python
b=a
```
- 메모리주소38에 존재하는 방을 b라고 하고, 그 방에 a를 넣어야하는데, a는 [1,2,3]이니까 [1,2,3]을 넣는다. 

(4) 아래를 실행하면 
```python
a.append(4)
```
- 방 a로가서 [1,2,3]을 [1,2,3,4]로 바꾼다. 
- 그리고 방 b에는 아무것도 하지 않는다. 

`-` R에서는 맞는 비유인데, 파이썬은 적절하지 않은 비유이다. 

In [11]:
id(a)

139851743661952

In [12]:
id(b)

139851743661952

실제로는 a,b가 저장된 메모리 주소가 동일함 

`-` 파이썬에서는 아래가 더 적절한 비유이다. 

(1) 메모리는 변수를 담을 방이 여러개 있는 호텔이라고 생각하자. 

(2) 아래를 실행하였을 경우 
```python
a=[1,2,3]
```
- 메모리주소139851743661952에서 [1,2,3]을 생성해요 
- 방 139851743661952의 방문에 a라는 포스트잇을 붙인다. 
- 앞으로 [1,2,3]에 접근하기 위해서는 여러 메모리방중에서 a라는 포스트잇이 붙은 방을 찾아가면 된다. 

(3) 아래를 실행하였을 경우 
```python
b=a
```
- a라는 포스트잇이 있는데, a라는 포스트잇이랑 b라는 포스트잇과 같은 효과를 주도록 한다. 
- 쉽게말하면 b라는 포스트잇을 방 139851743661952의 방문에 붙인다는 이야기. 
- 앞으로 [1,2,3]에 접근하기 위해서는 여러 메모리방중에서 a라는 포스트잇이 붙어 있거나 b라는 포스트잇이 붙어있는 방을 찾아가면 된다. 

(4) 아래를 실행하면 
```python
a.append(4)
```
- a라는 포스트잇이 붙어있는 방으로 가서, 그 내용물 append함수를 써서 4를 추가하라. 즉 내용물 [1,2,3]을 [1,2,3,4]로 바꾸라. 
- 같은방에 a,b라는 포스트잇이 모두 붙어있음. 따라서 b라는 포스트잇이 붙은 방을 찾아가서 내용물을 열어보면 [1,2,3,4]가 나온다. 

### 할당문(=)의 이해 

`-` 파이썬에서 할당문을 이해하기 위해서는 언제나 오른쪽을 먼저 읽어야 한다. 
- 할당문의 오른쪽에서는 객체를 생성하거나 가져옴
- 그 후에 라벨을 붙이듯이 할당문 왼쪽의 변수가 할당문 오른쪽의 객체에 바인딩 된다. (참조)

`-` `b=a`는 

> 나는 이미 a가 의미하는게 무엇인지 알고있어. 그런데 그 실체를 b라고도 부르고 싶어. 

라는 것과 같다. 즉 이미 a라고 부르고 있는것을 내가 b라고도 부르고싶다는 의미인데 이는 마치 별명과 같다. (b는 a의 별명, alias) 그리고 이처럼 하나의 오브젝트에 여러개의 이름을 붙이는 것을 에일리어싱이라고 부른다. 